src/lib/gssapi/krb5/set_context_flags.c

```diff
@@ -0,0 +1,41 @@
+/* -*- mode: c; c-basic-offset: 4; indent-tabs-mode: nil -*- */
+
+#include "k5-int.h"
+#include "gssapiP_krb5.h"
+
+#include <stdlib.h>
+
+OM_uint32 KRB5_CALLCONV
+krb5_gss_set_context_flags(OM_uint32 *minor_status, gss_ctx_id_t context,
+                           uint64_t req_flags, uint64_t ret_flags)
+{
+    krb5_gss_ctx_id_t external_context;
+
+    if (context == GSS_C_NO_CONTEXT)
+        return GSS_S_FAILURE | GSS_S_NO_CONTEXT;
+
+    if (minor_status != NULL)
+        *minor_status = 0;
+
+    external_context = (krb5_gss_ctx_id_t)context;
+    if (external_context->magic != KG_CONTEXT)
+        return GSS_S_FAILURE | GSS_S_NO_CONTEXT;
+
+    external_context->gss_flags = req_flags & (GSS_C_CONF_FLAG |
+                                               GSS_C_INTEG_FLAG |
+                                               GSS_C_MUTUAL_FLAG |
+                                               GSS_C_REPLAY_FLAG |
+                                               GSS_C_SEQUENCE_FLAG |
+                                               GSS_C_DELEG_FLAG |
+                                               GSS_C_DCE_STYLE |
+                                               GSS_C_IDENTIFY_FLAG |
+                                               GSS_C_EXTENDED_ERROR_FLAG);
+    external_context->gss_flags |= GSS_C_TRANS_FLAG;
+    if (req_flags & GSS_C_DCE_STYLE)
+        external_context->gss_flags |= GSS_C_MUTUAL_FLAG;
+
+    external_context->req_flags = req_flags;
```

Comment:
I think this is backwards.  You want to set `external_context->req_flags` and _not_ `external_context->gss_flags` here.  Before this change `external_context->gss_flags` was just `ret_flags`, and in this change it should be renamed to `external_context->ret_flags`, but in no case should this function modify the `ret_flags`.  The `ret_flags` can only be computed at accept-time -- they must not be settable by any setter.

The rule in GSS is that a) all features are optional (except channel binding, which is what we're fixing here!), b) you can get more features than you've asked for (e.g., you may not have asked for replay-detection, but Kerberos providers always give you that even though they don't have to if you didn't ask for it).

BTW, (b) means that ret_flags can be different on the initiator and acceptor side.